### DISCUSSION TITLE
Generating smiles instead of images of fingerprint bits

### DISCUSSION THREAD

**Discussion by ThomasWolf0701** [rdkit/rdkit#3579] [Q&A]
When looking at the documentation it is quite easy to generate images of substructures that set the bits of a Morgan Fingerprint. The description of extracting smiles instead of images seem far more complex and is decribed quite differently from generating images. Is there an easy way to get smiles that represent the images ?

**Comment by greglandrum** [comment]
@ThomasWolf0701 sorry for the slow reply on this one. I need to get used to looking at the discussions.

Are you looking for something beyond what's described here in the documentation?
https://www.rdkit.org/docs/GettingStartedInPython.html#explaining-bits-from-morgan-fingerprints
Specifically the last part of that section:
```
>>> atoms=set()
>>> for bidx in env:
...     atoms.add(m.GetBondWithIdx(bidx).GetBeginAtomIdx())
...     atoms.add(m.GetBondWithIdx(bidx).GetEndAtomIdx())
...
>>> Chem.MolFragmentToSmiles(m,atomsToUse=list(atoms),bondsToUse=env,rootedAtAtom=5)
'c(C)(cc)nc'
```

**Reply by miketynes** [reply]
Hi @greglandrum. I for one find this example very helpful. Is it similarly straightforward to generate a SMILES for a bit in an RDKitFingerprint? This would be a helpful example to add the docs!

**Reply by greglandrum** [reply]
@miketynes, it's more or less the same process for the RDKit fingerprint:
```
>>> m = Chem.MolFromSmiles('CCOc1ncccc1')
>>> bitinfo = {}
>>> fp = Chem.RDKFingerprint(m,bitInfo=bitinfo)
>>> bitinfo.keys()
dict_keys([0, 5, 11, 22, 44, 49, 72, 79, 103, 104, 124, 132, 154, 160, 161, 194, 208, 215, 242, 275, 280, 294, 305, 309, 330, 338, 339, 358, 361, 414, 430, 445, 472, 482, 496, 502, 521, 534, 549, 557, 560, 562, 576, 581, 596, 599, 608, 633, 650, 659, 665, 672, 684, 702, 723, 742, 777, 789, 794, 803, 804, 811, 829, 838, 842, 849, 866, 869, 879, 894, 896, 913, 914, 935, 943, 946, 959, 977, 1026, 1027, 1060, 1084, 1088, 1098, 1116, 1117, 1144, 1158, 1176, 1183, 1186, 1204, 1240, 1241, 1250, 1261, 1265, 1284, 1306, 1308, 1318, 1339, 1342, 1345, 1371, 1379, 1400, 1410, 1415, 1430, 1438, 1456, 1460, 1464, 1493, 1501, 1522, 1529, 1536, 1539, 1567, 1572, 1576, 1587, 1600, 1647, 1656, 1659, 1670, 1671, 1680, 1681, 1692, 1694, 1695, 1698, 1700, 1728, 1737, 1745, 1772, 1784, 1793, 1805, 1837, 1974, 1976, 2002, 2018, 2025, 2041])
>>> bitinfo[5]
[[3, 4]]
>>> atoms = set()
>>> for bidx in bitinfo[5][0]:
...     atoms.add(m.GetBondWithIdx(bidx).GetBeginAtomIdx())
...     atoms.add(m.GetBondWithIdx(bidx).GetEndAtomIdx())
... 
>>> Chem.MolFragmentToSmiles(m,atomsToUse=list(atoms),bondsToUse=bitinfo[5][0])
'cnc'
```

And I agree that something like this would be great for the Cookbook. Pull requests are always welcome! :-)
